File: templates/backends/python-fastapi/app/middleware/rate_limiter.py
from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field

from fastapi import HTTPException, Request


@dataclass
class Visitor:
    first_seen: float
    count: int = 1
    blocked: bool = False
    blocked_at: float = 0.0


@dataclass
class BlockedIP:
    ip: str
    blocked_at: float


class RateLimiter:
    def __init__(self, max_requests: int = 100, window_seconds: float = 60, block_seconds: float = 300):
        self._max_requests = max_requests
        self._window = window_seconds
        self._block_duration = block_seconds
        self._visitors: dict[str, Visitor] = {}
        self._cleanup_task: asyncio.Task[None] | None = None

    def start_cleanup(self) -> None:
        self._cleanup_task = asyncio.create_task(self._cleanup_loop())

    async def _cleanup_loop(self) -> None:
        while True:
            await asyncio.sleep(300)
            now = time.monotonic()
            expired = [
                ip
                for ip, v in self._visitors.items()
                if not v.blocked and (now - v.first_seen) > self._window
            ]
            for ip in expired:
                del self._visitors[ip]

    async def __call__(self, request: Request) -> None:
        ip = request.client.host if request.client else "unknown"
        if ip.startswith("::") or ip == "127.0.0.1":
            return

        now = time.monotonic()
        v = self._visitors.get(ip)

        if v is None:
            self._visitors[ip] = Visitor(first_seen=now)
            return

        if v.blocked:
            if (now - v.blocked_at) > self._block_duration:
                v.blocked = False
                v.count = 1
                v.first_seen = now
                return
            raise HTTPException(
                status_code=429,
                detail={"error": "Too many requests. You are temporarily blocked.", "error_code": "BLOCKED"},
            )

        if (now - v.first_seen) > self._window:
            v.count = 1
            v.first_seen = now
            return

        v.count += 1
        if v.count > self._max_requests:
            v.blocked = True
            v.blocked_at = now
            raise HTTPException(
                status_code=429,
                detail={"error": "Too many requests. You are temporarily blocked.", "error_code": "BLOCKED"},
            )

    def get_blocked_ips(self) -> list[BlockedIP]:
        return [
            BlockedIP(ip=ip, blocked_at=v.blocked_at)
            for ip, v in self._visitors.items()
            if v.blocked
        ]

    def unblock_ip(self, ip: str) -> bool:
        v = self._visitors.get(ip)
        if v is None or not v.blocked:
            return False
        v.blocked = False
        v.count = 0
        v.first_seen = time.monotonic()
        return True

File: backend/go-api/internal/middleware/ratelimiter.go
package middleware

import (
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

type visitor struct {
	firstSeen time.Time
	count     int
	blocked   bool
	blockedAt time.Time
}

type BlockedIP struct {
	IP        string    `json:"ip"`
	BlockedAt time.Time `json:"blocked_at"`
}

var (
	visitors = make(map[string]*visitor)
	mu       sync.Mutex
)

func RateLimiter(maxRequests int, window, blockDuration time.Duration) gin.HandlerFunc {
	go cleanupVisitors(window)

	return func(c *gin.Context) {
		ip := c.ClientIP()
		if strings.HasPrefix(ip, "::") || ip == "127.0.0.1" {
			c.Next()
			return
		}

		mu.Lock()
		v, exists := visitors[ip]

		if !exists {
			visitors[ip] = &visitor{firstSeen: time.Now(), count: 1}
			mu.Unlock()
			c.Next()
			return
		}

		if v.blocked {
			if time.Since(v.blockedAt) > blockDuration {
				v.blocked = false
				v.count = 1
				v.firstSeen = time.Now()
				mu.Unlock()
				c.Next()
				return
			}
			mu.Unlock()
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error":      "Too many requests. You are temporarily blocked.",
				"error_code": "BLOCKED",
			})
			return
		}

		if time.Since(v.firstSeen) > window {
			v.count = 1
			v.firstSeen = time.Now()
			mu.Unlock()
			c.Next()
			return
		}

		v.count++
		if v.count > maxRequests {
			v.blocked = true
			v.blockedAt = time.Now()
			mu.Unlock()
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error":      "Too many requests. You are temporarily blocked.",
				"error_code": "BLOCKED",
			})
			return
		}

		mu.Unlock()
		c.Next()
	}
}

func cleanupVisitors(window time.Duration) {
	for {
		time.Sleep(5 * time.Minute)
		mu.Lock()
		for ip, v := range visitors {
			if !v.blocked && time.Since(v.firstSeen) > window {
				delete(visitors, ip)
			}
		}
		mu.Unlock()
	}
}

func GetAllBlockedIPs() []BlockedIP {
	mu.Lock()
	defer mu.Unlock()

	var blocked []BlockedIP
	for ip, v := range visitors {
		if v.blocked {
			blocked = append(blocked, BlockedIP{IP: ip, BlockedAt: v.blockedAt})
		}
	}
	return blocked
}

func UnblockIPAddr(ip string) bool {
	mu.Lock()
	defer mu.Unlock()

	v, exists := visitors[ip]
	if !exists || !v.blocked {
		return false
	}

	v.blocked = false
	v.count = 0
	v.firstSeen = time.Now()
	return true
}
